src/main/java/org/testng/IExecutorFactory.java

```diff
@@ -0,0 +1,71 @@
+package org.testng;
+
+import java.util.Comparator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.testng.internal.DynamicGraph;
+import org.testng.internal.thread.graph.IThreadWorkerFactory;
```

Comment:
Something is wrong here. The public api must not expose the internal api otherwise the internal api is public

Reply:
True and agreed. But we are exposing an interface here and not an implementation. Without the factory being exposed (which is how we are creating threads) we can't get this done. I am open to suggestions.

Reply:
I suppose it should be possible with a deeper refactor and some inversion control.

Maybe a `ITestNGThreadPoolExecutor#run(List<ITestNGMethod> methods)` could be possible.

The idea is to rethink a bit the current flow instead of matching it exactly.

Reply:
@juherr  

> The idea is to rethink a bit the current flow instead of matching it exactly.

- Sure. But here's the problem that I am having. The current implementation requires:

* `org.testng.internal.DynamicGraph`
* `org.testng.internal.thread.graph.IThreadWorkerFactory`
* `org.testng.internal.thread.graph.IWorker`

for its working. 

I am not sure if its really worth the effort to decouple these from `org.testng.internal.thread.graph.GraphThreadPoolExecutor` just to support customized injection of thread pool executors. 

I was instead trying to see what is the minimal set of code changes that we can do to get this done. So with that in mind 

1. We can build an interface around `DynamicGraph` and expose that interface.
2. With respect to `IThreadWorkerFactory` and `IWorker` I think its better we expose these interfaces (since they are only interfaces and aren't exposing any functionality). 

That would basically help us implement this functionality with the minimal set of changes. 

Thoughts ?

Reply:
I like the minimal changes too but I've no idea how much effort is needed to make a nice refactoring. My feeling is it should not be too difficult but I can be totally wrong.
You can see it has a good design exercise too.

BTW, I let you choose the best solution and new interfaces is a good solution too :)

Reply:
Ok. Then let me go ahead and update my PR with what I suggested. You can take a look at it and let me know if that makes sense. I agree with you completely that this is a good design exercise. Let me see what I can do.